File: web-reference/client/src/lib/weatherIcons.ts
import type { WeatherCondition } from "@shared/schema";

/**
 * Icon logic — pure, side-effect free, easy to test.
 *
 * Maps a weather condition + day/night flag to a stable icon key. The key is
 * resolved to an actual SVG component in `@/components/weather-icon`. Keeping the
 * mapping here (logic) separate from the SVG rendering (component) lets us cache
 * and reason about icon decisions without touching React.
 */

export type WeatherIconKey =
  | "clear-day"
  | "clear-night"
  | "cloudy-day"
  | "cloudy-night"
  | "rainy-day"
  | "rainy-night"
  | "stormy"
  | "snowy"
  | "foggy"
  | "windy"
  | "hot"
  | "cold";

/**
 * Conditions whose appearance meaningfully changes with the sun being up or
 * down (sky-based). Temperature/air conditions (hot, cold, windy) and conditions
 * dominated by dark clouds (stormy, snowy, foggy) read the same day or night, so
 * they intentionally ignore `isDay`.
 */
const DAY_NIGHT_AWARE: ReadonlySet<WeatherCondition> = new Set<WeatherCondition>([
  "clear",
  "cloudy",
  "rainy",
]);

/**
 * Resolve the icon key for a condition and time of day.
 *
 * @param condition normalized weather condition from the API
 * @param isDay     true when the sun is above the horizon at the location
 */
export function getWeatherIcon(
  condition: WeatherCondition,
  isDay: boolean,
): WeatherIconKey {
  if (DAY_NIGHT_AWARE.has(condition)) {
    return `${condition}-${isDay ? "day" : "night"}` as WeatherIconKey;
  }
  return condition as WeatherIconKey;
}

/** Fallback day/night detection when the API gives us nothing (06:00–18:00 = day). */
export function isDayFromHour(hour: number): boolean {
  return hour >= 6 && hour < 18;
}

File: android/app/src/main/java/com/weatherquips/app/domain/model/WeatherIcons.kt
package com.weatherquips.app.domain.model

/**
 * Icon logic — pure and side-effect free, ported from `lib/weatherIcons.ts`.
 *
 * Only sky-based conditions change with day/night; temperature and air
 * conditions (hot, cold, windy) and dark-cloud ones (stormy, snowy, foggy) read
 * the same either way, so they deliberately ignore `isDay`.
 */
enum class WeatherIconKey(val id: String) {
    CLEAR_DAY("clear-day"),
    CLEAR_NIGHT("clear-night"),
    CLOUDY_DAY("cloudy-day"),
    CLOUDY_NIGHT("cloudy-night"),
    RAINY_DAY("rainy-day"),
    RAINY_NIGHT("rainy-night"),
    STORMY("stormy"),
    STORMY_NIGHT("stormy-night"),
    SNOWY("snowy"),
    SNOWY_NIGHT("snowy-night"),
    FOGGY("foggy"),
    WINDY("windy"),
    HOT("hot"),
    COLD("cold"),
}

/**
 * Conditions whose icon changes with the sun.
 *
 * The rule is whether the sky itself is part of the picture. Storms and snow
 * fall out of a sky you can still see, so they get a moon after dark. Fog
 * replaces the sky rather than sitting under it, and wind, heat and cold are
 * about the air rather than the sky at all — those look the same at any hour.
 */
private val DAY_NIGHT_AWARE = setOf(
    WeatherCondition.CLEAR,
    WeatherCondition.CLOUDY,
    WeatherCondition.RAINY,
    WeatherCondition.STORMY,
    WeatherCondition.SNOWY,
)

fun weatherIconKey(condition: WeatherCondition, isDay: Boolean): WeatherIconKey =
    if (condition in DAY_NIGHT_AWARE) {
        when (condition) {
            WeatherCondition.CLEAR ->
                if (isDay) WeatherIconKey.CLEAR_DAY else WeatherIconKey.CLEAR_NIGHT
            WeatherCondition.CLOUDY ->
                if (isDay) WeatherIconKey.CLOUDY_DAY else WeatherIconKey.CLOUDY_NIGHT
            WeatherCondition.STORMY ->
                if (isDay) WeatherIconKey.STORMY else WeatherIconKey.STORMY_NIGHT
            WeatherCondition.SNOWY ->
                if (isDay) WeatherIconKey.SNOWY else WeatherIconKey.SNOWY_NIGHT
            else ->
                if (isDay) WeatherIconKey.RAINY_DAY else WeatherIconKey.RAINY_NIGHT
        }
    } else {
        when (condition) {
            WeatherCondition.STORMY -> WeatherIconKey.STORMY
            WeatherCondition.SNOWY -> WeatherIconKey.SNOWY
            WeatherCondition.FOGGY -> WeatherIconKey.FOGGY
            WeatherCondition.WINDY -> WeatherIconKey.WINDY
            WeatherCondition.HOT -> WeatherIconKey.HOT
            else -> WeatherIconKey.COLD
        }
    }

/** Fallback day/night detection when a provider gives us nothing (06:00–18:00 = day). */
fun isDayFromHour(hour: Int): Boolean = hour in 6..17
